### PR TITLE
Fix some pyright member access errors in training module

### LINKED ISSUE #2093: CI: Add type checking
Related to #855 add type check in CI

### PR DESCRIPTION
**Pull request by uditarora**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## What does this PR do?
Part of #855.
Fixes #2093

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by uditarora** [comment]
I'm looking at the failing tests as a result of this change. I'll update once this is ready to review.

**Comment by uditarora** [comment]
> mind add testing of these fixes?

I'm not sure what kind of testing you are referring to. Do you want me to add new tests to check the specified types?

**Comment by Borda** [comment]
> I'm not sure what kind of testing you are referring to. Do you want me to add new tests to check the specified types?

add a GH action, see https://medium.com/@doedotdev/mypy-for-github-action-7da1ebee99e7

**Comment by uditarora** [comment]
@Borda I think I might have messed up the version history a bit by rebasing the master onto this branch instead of merging. Should I continue with this branch or just start fresh with a new branch and a new PR?

**Comment by Borda** [comment]
> @Borda I think I might have messed up the version history a bit by rebasing the master onto this branch instead of merging. Should I continue with this branch or just start fresh with a new branch and a new PR?

let me check it for you...

**Comment by Borda** [comment]
@uditarora seems to be fixed now

**Comment by uditarora** [comment]
I've added a list of files/directories to be ignored by pyright in the pyrightconfig.json file. When more files are fixed, this list can be updated without changing anything in the workflow.

**Comment by uditarora** [comment]
The failing check seems to be caused by a failure while trying to upload pytest results on one config, unrelated to code changes.

File: pyrightconfig.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{
```
Comment (Borda) [inline pyrightconfig.json:1]: could we move this to `tests` folder?

File: .github/workflows/python-type-check.yml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+          # python -m pip install --upgrade --user pip
+          pip install -r requirements.txt -U -f https://download.pytorch.org/whl/torch_stable.html -q
+          HOROVOD_BUILD_ARCH_FLAGS="-mfma" pip install -r ./tests/requirements-devel.txt -q
+          # pip install tox coverage
+          python --version
+          pip --version
+          pip list
+        shell: bash
+
+      - name: Reinstall Horovod if necessary
```
Comment (Borda) [inline .github/workflows/python-type-check.yml:59]: I ques we can skip this as the Horovod is not trully launched

**Comment by uditarora** [comment]
Thank you for adding the suggested changes!

I'll keep fixing more pyright errors as and when I get time, and send a PR once I have a bunch of them.

**Comment by Borda** [comment]
@uditarora GREAT work! happy to see more or this fixes... :]

**Pull request merged**